File: docs/core/examples/ruby/control_programs.rb
require 'chain'

chain = Chain::Client.new
signer = Chain::HSMSigner.new

key = chain.mock_hsm.keys.create
signer.add_key(key, chain.mock_hsm.signer_conn)

chain.assets.create(
  alias: 'gold',
  root_xpubs: [key.xpub],
  quorum: 1,
)

chain.accounts.create(
  alias: 'alice',
  root_xpubs: [key.xpub],
  quorum: 1,
)

chain.accounts.create(
  alias: 'bob',
  root_xpubs: [key.xpub],
  quorum: 1,
)

tx = chain.transactions.build do |b|
  b.issue asset_alias: 'gold', amount: 100
  b.control_with_account account_alias: 'bob', asset_alias: 'gold', amount: 100
end

chain.transactions.submit(signer.sign(tx))

# snippet create-receiver
alice_receiver = chain.accounts.create_receiver(
  account_alias: 'alice'
)
alice_receiver_serialized = alice_receiver.to_json
# endsnippet

# snippet build-transaction
payment_to_receiver = chain.transactions.build do |b|
  b.spend_from_account account_alias: 'bob', asset_alias: 'gold', amount: 10
  b.control_with_receiver(
    receiver: JSON.parse(alice_receiver_serialized),
    asset_alias: 'gold',
    amount: 10
  )
end

chain.transactions.submit(signer.sign(payment_to_receiver))
# endsnippet

# snippet retire
retirement = chain.transactions.build do |b|
  b.spend_from_account account_alias: 'alice', asset_alias: 'gold', amount: 10
  b.retire asset_alias: 'gold', amount: 10
end

chain.transactions.submit(signer.sign(retirement))
# endsnippet

File: docs/core/examples/node/controlPrograms.js
const chain = require('chain-sdk')

const client = new chain.Client()
const signer = new chain.HsmSigner()

client.mockHsm.keys.create().then(key => {
  signer.addKey(key.xpub, client.mockHsm.signerConnection)
  return key.xpub
}).then(xpub =>
  Promise.all([
    client.assets.create({
      alias: 'gold',
      rootXpubs: [xpub],
      quorum: 1,
    }),
    client.accounts.create({
      alias: 'alice',
      rootXpubs: [xpub],
      quorum: 1
    }),
    client.accounts.create({
      alias: 'bob',
      rootXpubs: [xpub],
      quorum: 1
    })
  ])
).then(() =>
  client.transactions.build(builder => {
    builder.issue({
      assetAlias: 'gold',
      amount: 100
    })
    builder.controlWithAccount({
      accountAlias: 'bob',
      assetAlias: 'gold',
      amount: 100
    })
  })
).then(
  template => signer.sign(template)
).then(
  signed => client.transactions.submit(signed)
).then(() => {
  // snippet create-control-program
  const aliceProgramPromise = client.accounts.createControlProgram({
    alias: 'alice',
  })
  // endsnippet

  return aliceProgramPromise
}).then(aliceProgram => {
  // snippet build-transaction
  return client.transactions.build(builder => {
    builder.spendFromAccount({
      accountAlias: 'bob',
      assetAlias: 'gold',
      amount: 10
    })
    builder.controlWithProgram({
      controlProgram: aliceProgram.controlProgram,
      assetAlias: 'gold',
      amount: 10
    })
  }).then(template => {
    return signer.sign(template)
  }).then(signed => {
    return client.transactions.submit(signed)
  })
  // endsnippet
}).then(() =>
  // snippet retire
  client.transactions.build(builder => {
    builder.spendFromAccount({
      accountAlias: 'alice',
      assetAlias: 'gold',
      amount: 10
    })
    builder.retire({
      assetAlias: 'gold',
      amount: 10
    })
  }).then(template => {
    return signer.sign(template)
  }).then(signed => {
    return client.transactions.submit(signed)
  })
  // endsnippet
).catch(err =>
  process.nextTick(() => { throw err })
)
